numba/unicode.py

```diff
@@ -1148,6 +1152,84 @@ def impl(a):
     return impl
 
 
+@register_jitable
+def _handle_capital_sigma(data, length, idx):
+    """
+    https://github.com/python/cpython/blob/3.6/Objects/unicodeobject.c#L9856-L9883
+    """
+    c = 0
+    j = idx - 1
+    for j in range(idx - 1, -1, -1):
+        c = _get_code_point(data, j)
+        if not _PyUnicode_IsCaseIgnorable(_Py_UCS4(c)):
+            break
+    final_sigma = (j >= 0 and _PyUnicode_IsCased(_Py_UCS4(c)))
+    if final_sigma:
+        for j in range(idx + 1, length):
+            c = _get_code_point(data, j)
+            if not _PyUnicode_IsCaseIgnorable(_Py_UCS4(c)):
+                break
+        final_sigma = (j == length or not _PyUnicode_IsCased(_Py_UCS4(c)))
+
+    return 0x3C2 if final_sigma else 0x3C3
+
+
+@register_jitable
+def _lower_ucs4(code_point, data, length, idx, mapped):
+    """
+    https://github.com/python/cpython/blob/3.6/Objects/unicodeobject.c#L9885-L9895
+    """
+    if code_point == 0x3A3:
+        mapped[0] = _handle_capital_sigma(data, length, idx)
+        return 1
+    return _PyUnicode_ToLowerFull(_Py_UCS4(code_point), mapped)
+
+
+@register_jitable
+def _do_title(data, length, res, maxchars):
+    """
+    https://github.com/python/cpython/blob/3.6/Objects/unicodeobject.c#L9996-L10021
+    """
+    k = 0
+    previous_cased = False
+    for idx in range(length):
+        mapped = np.zeros(3, dtype=_Py_UCS4)
+        code_point = _get_code_point(data, idx)
+        if previous_cased:
+            n_res = _lower_ucs4(code_point, data, length, idx, mapped)
+        else:
+            n_res = _PyUnicode_ToTitleFull(_Py_UCS4(code_point), mapped)
+        for m in mapped[:n_res]:
+            maxchar, = maxchars
+            maxchars[0] = max(maxchar, m)
+            _set_code_point(res, k, m)
+            k += 1
+        previous_cased = _PyUnicode_IsCased(_Py_UCS4(code_point))
+    return k
+
+
+@overload_method(types.UnicodeType, 'title')
+def unicode_title(data):
+    """
+    Implements .title()
+    https://github.com/python/cpython/blob/3.6/Objects/unicodeobject.c#L10023-L10069
```

Comment:
Suggest to describe here how the algorithm works.

Reply:
Added.